addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerThingHandler.java

```diff
@@ -0,0 +1,359 @@
+                // mgb: call getFoo() for the prefix 'foo'
+                // - mac -> getMac()
+                // - ip -> getIp()
+                // ...etc...
+                String suffix = invokeMethod(client, "get" + StringUtils.capitalize(prefix));
+                if (StringUtils.isNotBlank(suffix)) {
+                    String key = prefix + CACHE_KEY_SEPARATOR + suffix;
+                    cache.put(key, client);
+                }
+            }
+        }
+    }
+
+    private @Nullable UniFiClient cacheGet(Map<String, UniFiClient> cache, String cid) {
+        UniFiClient client = null;
+        synchronized (cache) {
+            for (String prefix : CACHE_KEY_PREFIXES) {
+                String key = prefix + CACHE_KEY_SEPARATOR + cid;
+                if (cache.containsKey(key)) {
+                    client = cache.get(key);
+                    logger.debug("Found client matching '{}'", key);
+                    logger.debug("  {}", client);
+                    break;
+                }
+            }
+        }
+        return client;
+    }
+
+    public @Nullable UniFiClient getClient(String cid, String site) {
+        // mgb: first check active clients and fallback to insights if not found
+        UniFiClient client = null;
+
+        // mgb: first check active clients and fallback to insights if not found
+        client = cacheGet(clientsCache, cid);
+        if (client == null) {
+            client = cacheGet(insightsCache, cid);
+        }
+
+        // mgb: short circuit
+        if (client == null || BooleanUtils.isNotTrue(client.isWireless()) || !belongsToSite(client, site)) {
+            return null;
+        }
+
+        // mgb: instanceof check just for type / cast safety
+        return (client instanceof UniFiWirelessClient ? (UniFiWirelessClient) client : null);
+    }
+
+    // Private API
+
+    private void scheduleRefreshJob() {
+        synchronized (this) {
+            if (refreshJob == null) {
+                logger.debug("Scheduling refresh job every {}s", config.getRefresh());
+                refreshJob = scheduler.scheduleWithFixedDelay(this::run, 0, config.getRefresh(), TimeUnit.SECONDS);
+            }
+        }
+    }
+
+    private void cancelRefreshJob() {
+        synchronized (this) {
+            if (refreshJob != null) {
+                logger.debug("Cancelling refresh job");
+                refreshJob.cancel(true);
+                refreshJob = null;
+            }
+        }
+    }
+
+    private void run() {
+        try {
+            logger.trace("Executing refresh job");
+            refresh();
+            updateStatus(ONLINE);
+        } catch (UniFiCommunicationException e) {
+            updateStatus(OFFLINE, COMMUNICATION_ERROR, STATUS_DESCRIPTION_COMMUNICATION_ERROR);
+        } catch (UniFiInvalidCredentialsException e) {
+            updateStatus(OFFLINE, CONFIGURATION_ERROR, STATUS_DESCRIPTION_INVALID_CREDENTIALS);
+        } catch (Exception e) {
```

Comment:
IIRC I left this as `Exception` because the `ScheduledExecutorService` will stop executing scheduled tasks if an exception is thrown. I'm not sure if this has since been fixed.

[Here was the previous discussion we had on this subject](https://github.com/openhab/openhab2-addons/pull/3387#discussion_r177454322)